crates/bevy_animation/src/curve_path.rs

```diff
@@ -0,0 +1,159 @@
+//! A curve through space.
+
+use bevy_math::cubic_splines::{CubicCurve, Point};
+
+/// A curve representing a path through space. Unlike [`CubicCurve`], it is sampled based on distance
+/// along the path. Can be used for linearly animating along a curve.
+///
+/// Note: Values are approximated using a series of line segments. Accuracy is based on the
+/// subdivisions specified when creating the path.
+#[derive(Clone, Debug)]
+pub struct CurvePath<P: Point> {
+    curve: CubicCurve<P>,
+    arc_lengths: Vec<f32>,
+}
+
+impl<P: Point> CurvePath<P> {
+    /// Creates a new [`CurvePath`] from a curve. Subdivisions will determine how many line segments
+    /// are used, and therefore the accuracy.
+    pub fn from_cubic_curve(curve: CubicCurve<P>, subdivisions: usize) -> CurvePath<P> {
+        let arc_lengths: Vec<f32> = curve
+            .iter_positions(subdivisions)
+            .scan((0.0, curve.position(0.0)), |state, x| {
+                state.0 += x.distance(state.1);
+                state.1 = x;
+
+                Some(state.0)
+            })
+            .collect();
+
+        CurvePath { curve, arc_lengths }
+    }
+
+    /// Returns the [`CubicCurve`] this path is based on.
+    pub fn curve(&self) -> &CubicCurve<P> {
+        &self.curve
+    }
+
+    /// Total length of the curve.
+    pub fn length(&self) -> f32 {
+        self.arc_lengths[self.arc_lengths.len() - 1]
+    }
+
+    /// Returns a 't' value corresponding to the `t` value on the underlying curve that matches the
+    /// provided `distance`.
+    pub fn t(&self, distance: f32) -> f32 {
+        // Get index with greatest value that is less than or equal to target distance.
+        let closest = self.arc_lengths.partition_point(|&x| x <= distance) - 1;
+
+        // Check if index's distance perfectly matches target distance, otherwise lerp between the
+        // index and its next neighbor.
+        if self.arc_lengths[closest] == distance {
+            (closest + 1) as f32 / (self.arc_lengths.len() + 1) as f32
+        } else {
+            let length_before = self.arc_lengths[closest];
+            let length_after = self.arc_lengths[closest + 1];
+            let segment_length = length_after - length_before;
+
+            let segment_fraction = (distance - length_before) / segment_length;
+
+            (closest as f32 + segment_fraction) / (self.arc_lengths.len() - 1) as f32
```

Comment:
I think this is wrong, or at least I don't see how this is right.

This:

```rust
(closest as f32 + segment_fraction) / (self.arc_lengths.len() - 1) as f32
```

is the normalized arc length in `[0:1]`. So dividing that by the number of cubic curve segments yields a correct value only if all segments of that cubic curve have the same arc length to begin with, which I don't see anywhere `CubicCurve` guaranteeing.

Reply:
I believe same-arc-length is not a requirement for this line to work. It's not working with distances but with indices, `closest` is the index of the division vertex before the desired point, `closest + 1` is the subdivision vertex after the desired point, `closest + segment_fraction` is the approximation of the "index" in the middle of this segment.

The "index" is not the physical length, it corresponds to the subdivisions.

One then needs to convert the index into a `t` parameter value, hence they divide by the number of subdividing segments in order to normalise the result into range <0; 1>. This is based on the assumption that all subdivisions are equally spaced in `t`.